File: app/src/main/java/com/angogasapps/myfamily/async/LoadFamilyThread.java
package com.angogasapps.myfamily.async;

import android.content.Context;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.AsyncTask;
import android.util.Log;

import androidx.annotation.NonNull;


import com.angogasapps.myfamily.firebase.FirebaseVarsAndConsts;
import com.angogasapps.myfamily.objects.User;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.ValueEventListener;


import java.io.InputStream;
import java.net.URL;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import static com.angogasapps.myfamily.firebase.FirebaseVarsAndConsts.CHILD_EMBLEM;
import static com.angogasapps.myfamily.firebase.FirebaseVarsAndConsts.CHILD_MEMBERS;
import static com.angogasapps.myfamily.firebase.FirebaseVarsAndConsts.CHILD_NAME;
import static com.angogasapps.myfamily.firebase.FirebaseVarsAndConsts.DATABASE_ROOT;

import static com.angogasapps.myfamily.firebase.FirebaseVarsAndConsts.DEFAULT_URL;
import static com.angogasapps.myfamily.firebase.FirebaseVarsAndConsts.NODE_FAMILIES;

import static com.angogasapps.myfamily.firebase.FirebaseVarsAndConsts.NODE_USERS;
import static com.angogasapps.myfamily.firebase.FirebaseVarsAndConsts.USER;
import static com.angogasapps.myfamily.firebase.FirebaseVarsAndConsts.familyMembersMap;
import static com.angogasapps.myfamily.utils.WithUsers.getUserFromSnapshot;


public final class LoadFamilyThread extends AsyncTask<User, Integer, ArrayList<User>> {
    public static volatile boolean isEnd = false;
    Context context;
    volatile ArrayList<String> familyMembersId = new ArrayList<>();
    volatile ArrayList<User> familyMembersList = new ArrayList<>();
    volatile HashMap<String, String> usersRoleMap = new HashMap<>();
    volatile String familyEmblemURL;
    volatile Bitmap familyEmblem;
    volatile String familyName;

    volatile ExecutorService pull;

//    volatile static Lock lock = new ReentrantLock();

    volatile static CountDownLatch lock = new CountDownLatch(1);

    public LoadFamilyThread(Context context) {
        this.context = context;

    }

    @Override
    protected ArrayList<User> doInBackground(User... users) {
        try {
            Log.v("tag", "Thread = " + Thread.currentThread());
            FamilyDownloaderThread familyDownloaderThread = new FamilyDownloaderThread();
            FamilyEmblemDownloaderThread emblemDownloaderThread = new FamilyEmblemDownloaderThread();

            familyDownloaderThread.start();
            familyDownloaderThread.join();

            downloadMembers();

            downloadMembersPhoto();

            ArrayList<User> a = familyMembersList;

            emblemDownloaderThread.start();

            initConnections();

            if (emblemDownloaderThread.isAlive())
                emblemDownloaderThread.join();

            Log.i("tag", "Загрузка завершена");

        } catch (Exception e) {
            e.printStackTrace();
        }
        HashMap<String, User> aStringUserHashMap = familyMembersMap;
        isEnd = true;
        context = null;
        return null;
    }

    private void downloadMembers(){
        pull = Executors.newFixedThreadPool(2);
        CountDownLatch monitor = new CountDownLatch(familyMembersId.size());
        for (int i = 0; i < familyMembersId.size(); i++) {
            pull.submit(new MemberDownloaderThread(familyMembersId.get(i), monitor));
        }
        pull.shutdown();
        try {
            monitor.await();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

    }


    private void downloadMembersPhoto(){
        pull = Executors.newFixedThreadPool(2);
        CountDownLatch monitor = new CountDownLatch(familyMembersList.size());
        for (int i = 0; i < familyMembersList.size(); i++) {
            pull.submit(new PhotoDownloaderThread(familyMembersList.get(i).getPhotoURL(), i, monitor));
        }
        pull.shutdown();
        try {
            monitor.await();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    private Bitmap downloadBitmapByURL(String url){
        try{
            URL photoUrl = new URL(url);
            InputStream downloadStream = photoUrl.openStream();
            Bitmap bitmap = BitmapFactory.decodeStream(downloadStream);
            return bitmap;
        }catch (Exception e) {
            if (url != null)
            if (!url.equals(DEFAULT_URL))
                e.printStackTrace();
        }
        return null;
    }


    private class FamilyDownloaderThread extends Thread {

        @Override
        public void run() {
            DATABASE_ROOT.child(NODE_FAMILIES).child(USER.getFamily())
                    .addListenerForSingleValueEvent(new ValueEventListener() {
                        @Override
                        public void onDataChange(@NonNull DataSnapshot snapshot) {
                            for (DataSnapshot dataSnapshot : snapshot.child(CHILD_MEMBERS).getChildren()) {
                                familyMembersId.add(dataSnapshot.getKey());
                                String string = dataSnapshot.getValue(String.class);
                                usersRoleMap.put(dataSnapshot.getKey(), string);
                            }
                            familyEmblemURL = snapshot.child(CHILD_EMBLEM).getValue(String.class);
                            familyName = snapshot.child(CHILD_NAME).getValue(String.class);

                            lock.countDown();
                        }

                        @Override
                        public void onCancelled(@NonNull DatabaseError error) {
                        }
                    });
            try {
                lock.await();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }

    }

    private class MemberDownloaderThread extends Thread{
        String memberId;
        CountDownLatch monitor;


        public MemberDownloaderThread(String id, CountDownLatch monitor){
            this.memberId = id;
            this.monitor = monitor;
        }

        @Override
        public void run() {
            DATABASE_ROOT.child(NODE_USERS).child(memberId)
                    .addListenerForSingleValueEvent(new ValueEventListener() {
                @Override
                public void onDataChange(@NonNull DataSnapshot snapshot) {
//                    familyMembersList.add(snapshot.getValue(User.class));
                    familyMembersList.add(getUserFromSnapshot(snapshot));
                    Log.v("tag", "Thread = " + Thread.currentThread());

                    MemberDownloaderThread.this.monitor.countDown();

                }

                @Override
                public void onCancelled(@NonNull DatabaseError error) {}
            });
        }
    }

    private class PhotoDownloaderThread extends Thread{
        String photoUrl;
        int pos;
        CountDownLatch monitor;

        public PhotoDownloaderThread(String photoUrl, int pos, CountDownLatch monitor) {
            this.photoUrl = photoUrl;
            this.pos = pos;
            this.monitor = monitor;
        }

        @Override
        public void run() {
            familyMembersList.get(pos).setBitmap(downloadBitmapByURL(this.photoUrl));
            PhotoDownloaderThread.this.monitor.countDown();
        }
    }

    private class FamilyEmblemDownloaderThread extends Thread{
        @Override
        public void run() {
            familyEmblem = downloadBitmapByURL(familyEmblemURL);
        }
    }

    private void initConnections(){
        FirebaseVarsAndConsts.familyEmblemImage = this.familyEmblem;
        for (User user : familyMembersList){
            user.setRole(usersRoleMap.get(user.getId()));
            familyMembersMap.put(user.getId(), user);
        }
    }
}

File: app/src/main/java/com/angogasapps/myfamily/async/LoadFamilyThread.kt
package com.angogasapps.myfamily.async

import android.content.Context
import com.angogasapps.myfamily.models.User.Companion.from

import android.os.AsyncTask
import kotlin.jvm.Volatile
import android.graphics.Bitmap
import com.angogasapps.myfamily.async.LoadFamilyThread.FamilyDownloaderThread
import com.angogasapps.myfamily.async.LoadFamilyThread.FamilyEmblemDownloaderThread
import com.angogasapps.myfamily.async.LoadFamilyThread
import com.angogasapps.myfamily.async.LoadFamilyThread.MemberDownloaderThread
import com.angogasapps.myfamily.async.LoadFamilyThread.PhotoDownloaderThread
import android.graphics.BitmapFactory
import android.util.Log
import com.google.firebase.database.ValueEventListener
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.angogasapps.myfamily.firebase.*
import com.angogasapps.myfamily.database.DatabaseManager
import com.angogasapps.myfamily.models.Family
import com.angogasapps.myfamily.models.User
import java.lang.Exception
import java.net.URL
import java.util.ArrayList
import java.util.HashMap
import java.util.concurrent.CountDownLatch
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

class LoadFamilyThread(private var context: Context) : AsyncTask<User?, Int?, ArrayList<User>?>() {
    @Volatile
    var familyMembersId = ArrayList<String>()

    @Volatile
    var familyMembersList = ArrayList<User>()

    @Volatile
    var usersRoleMap: HashMap<String, String> = HashMap()

    @Volatile
    var familyEmblemURL: String? = null

    @Volatile
    var familyEmblem: Bitmap? = null

    @Volatile
    lateinit var familyName: String

    @Volatile
    lateinit var pull: ExecutorService

    protected override fun doInBackground(vararg p0: User?): ArrayList<User>? {
        try {
            Log.v("tag", "Thread = " + Thread.currentThread())
            val familyDownloaderThread = FamilyDownloaderThread()
            val emblemDownloaderThread = FamilyEmblemDownloaderThread()
            familyDownloaderThread.start()
            familyDownloaderThread.join()
            downloadMembers()
            initUsersArray()
            downloadMembersPhoto()

//            ArrayList<User> a = familyMembersList;
            emblemDownloaderThread.start()
            initPhotos()
            if (emblemDownloaderThread.isAlive) emblemDownloaderThread.join()
            Log.i("load", "Загрузка завершена")
        } catch (e: Exception) {
            e.printStackTrace()
        }
        isEnd = true
        return null
    }

    private fun downloadMembers() {
        pull = Executors.newFixedThreadPool(2)
        val monitor = CountDownLatch(familyMembersId.size)
        for (i in familyMembersId.indices) {
            pull.submit(MemberDownloaderThread(familyMembersId[i], monitor))
        }
        pull.shutdown()
        try {
            monitor.await()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
    }

    private fun downloadMembersPhoto() {
        pull = Executors.newFixedThreadPool(2)
        val monitor = CountDownLatch(familyMembersList.size)
        for (i in familyMembersList.indices) {
            pull.submit(PhotoDownloaderThread(familyMembersList[i].photoURL, i, monitor))
        }
        pull.shutdown()
        try {
            monitor.await()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
    }

    private fun downloadBitmapByURL(url: String?): Bitmap? {
        try {
            val photoUrl = URL(url)
            val downloadStream = photoUrl.openStream()
            return BitmapFactory.decodeStream(downloadStream)
        } catch (e: Exception) {
            if (url != null) if (url != DEFAULT_URL) e.printStackTrace()
        }
        return null
    }

    private inner class FamilyDownloaderThread : Thread() {
        override fun run() {
            DATABASE_ROOT.child(NODE_FAMILIES).child(USER.family)
                .addListenerForSingleValueEvent(object : ValueEventListener {
                    override fun onDataChange(snapshot: DataSnapshot) {
                        for (dataSnapshot in snapshot.child(CHILD_MEMBERS).children) {
                            familyMembersId.add(dataSnapshot.key!!)
                            val string = dataSnapshot.getValue(String::class.java)!!
                            usersRoleMap[dataSnapshot.key!!] = string
                        }
                        familyEmblemURL = snapshot.child(CHILD_EMBLEM).getValue(String::class.java)
                        familyName = snapshot.child(CHILD_NAME).getValue(String::class.java)!!
                        lock.countDown()
                    }

                    override fun onCancelled(error: DatabaseError) {}
                })
            try {
                lock.await()
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
        }
    }

    private inner class MemberDownloaderThread(var memberId: String, var monitor: CountDownLatch) :
        Thread() {
        override fun run() {
            DATABASE_ROOT.child(NODE_USERS).child(memberId)
                .addListenerForSingleValueEvent(object : ValueEventListener {
                    override fun onDataChange(snapshot: DataSnapshot) {
//                    familyMembersList.add(snapshot.getValue(User.class));
                        familyMembersList.add(from(snapshot))
                        Log.v("tag", "Thread = " + currentThread())
                        monitor.countDown()
                    }

                    override fun onCancelled(error: DatabaseError) {}
                })
        }
    }

    private inner class PhotoDownloaderThread(
        var photoUrl: String,
        var pos: Int,
        var monitor: CountDownLatch
    ) : Thread() {
        override fun run() {
            familyMembersList[pos].setBitmap(downloadBitmapByURL(photoUrl))
            monitor.countDown()
        }
    }

    private inner class FamilyEmblemDownloaderThread : Thread() {
        override fun run() {
            familyEmblem = downloadBitmapByURL(familyEmblemURL)
        }
    }

    private fun initPhotos() {
        familyEmblemImage = familyEmblem
        for (user in familyMembersList) {
            user.role = usersRoleMap[user.id]!!
        }
        //        Family.getInstance().setUsersList(familyMembersList);
        DatabaseManager.updateInfoForUsers(familyMembersList)
    }

    private fun initUsersArray() {
        Family.getInstance().usersList = familyMembersList
    }

    companion object {
        @Volatile
        var isEnd = false
        //    volatile static Lock lock = new ReentrantLock();
        @Volatile
        private var lock = CountDownLatch(1)
    }
}